tokio/src/sync/broadcast.rs

```diff
@@ -0,0 +1,980 @@
+    /// ```
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx1) = broadcast::channel(16);
+    ///     let mut rx2 = tx.subscribe();
+    ///
+    ///     tokio::spawn(async move {
+    ///         assert_eq!(rx1.recv().await.unwrap(), 10);
+    ///         assert_eq!(rx1.recv().await.unwrap(), 20);
+    ///     });
+    ///
+    ///     tokio::spawn(async move {
+    ///         assert_eq!(rx2.recv().await.unwrap(), 10);
+    ///         assert_eq!(rx2.recv().await.unwrap(), 20);
+    ///     });
+    ///
+    ///     tx.send(10).unwrap();
+    ///     tx.send(20).unwrap();
+    /// }
+    /// ```
+    ///
+    /// Handling lag
+    ///
+    /// ```
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = broadcast::channel(2);
+    ///
+    ///     tx.send(10).unwrap();
+    ///     tx.send(20).unwrap();
+    ///     tx.send(30).unwrap();
+    ///
+    ///     // The receiver lagged behind
+    ///     assert!(rx.recv().await.is_err());
+    ///
+    ///     // At this point, we can abort or continue with lost messages
+    ///
+    ///     assert_eq!(20, rx.recv().await.unwrap());
+    ///     assert_eq!(30, rx.recv().await.unwrap());
+    /// }
+    pub async fn recv(&mut self) -> Result<T, RecvError> {
+        use crate::future::poll_fn;
+
+        poll_fn(|cx| self.poll_recv(cx)).await
+    }
+
+    fn register_waker(&self, cx: &Waker) {
+        self.wait.waker.register_by_ref(cx);
+
+        if !self.wait.queued.load(SeqCst) {
+            // Set `queued` before queuing.
+            self.wait.queued.store(true, SeqCst);
+
+            let mut curr = self.shared.wait_stack.load(SeqCst);
+            let node = Arc::into_raw(self.wait.clone()) as *mut _;
+
+            loop {
+                // Safety: `queued == false` means the caller has exclusive
+                // access to `self.wait.next`.
+                self.wait.next.with_mut(|ptr| unsafe { *ptr = curr });
+
+                let res = self.shared.wait_stack.compare_exchange(curr, node, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return,
+                    Err(actual) => curr = actual,
+                }
+            }
+        }
+    }
+}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        let mut tail = self.shared.tail.lock().unwrap();
+
+        tail.rx_cnt -= 1;
+        let until = tail.pos;
+
+        drop(tail);
+
+        while self.next != until {
+            match self.recv_ref(true) {
+                // Ignore the value
+                Ok(_) => {}
+                // The channel is closed
+                Err(TryRecvError::Closed) => break,
+                // Ignore lagging, we will catch up
+                Err(TryRecvError::Lagged) => {}
+                // Can't be empty
+                Err(TryRecvError::Empty) => panic!("unexpected empty broadcast channel"),
```

Comment:
This is a drop impl, should we check if the thread is already panicking here?

Reply:
I'd rather abort... if the panic is hit it is a severe internal bug.